devito/ir/clusters/algorithms.py

```diff
@@ -432,6 +443,57 @@ def callback(self, clusters, prefix, seen=None):
         return processed
 
 
+def reduction_comms(clusters):
+    processed = []
+    fifo = []
+    for c in clusters:
+        # Schedule the global distributed reductions encountered before `c`,
+        # if `c`'s IterationSpace is such that the reduction can be carried out
+        found, fifo = split(fifo, lambda dr: dr.ispace.is_subset(c.ispace))
+        if found:
+            exprs = [Eq(dr.var, dr) for dr in found]
+            processed.append(c.rebuild(exprs=exprs))
+
+        # Detect the global distributed reductions in `c`
+        for e in c.exprs:
+            op = e.operation
+            if op is None or c.is_sparse:
+                continue
+
+            var = e.lhs
+            grid = c.grid
+            if grid is None:
+                continue
+
+            # Is Inc/Max/Min/... actually used for a reduction?
+            ispace = c.ispace.project(lambda d: d in var.free_symbols)
+            if ispace.itdims == c.ispace.itdims:
+                continue
+
+            # The reduced Dimensions
+            rdims = set(c.ispace.itdims) - set(ispace.itdims)
+
+            # The reduced Dimensions inducing a global distributed reduction
+            grdims = {d for d in rdims if d._defines & c.dist_dimensions}
+            if not grdims:
+                continue
+
+            # The IterationSpace within which the global distributed reduction
+            # must be carried out
+            ispace = c.ispace.prefix(lambda d: d in var.free_symbols)
+
+            fifo.append(DistReduce(var, op=op, grid=grid, ispace=ispace))
+
+        processed.append(c)
+
+    # Leftover reductions are placed at the very end
+    if fifo:
```

Comment:
An example of what goes here?

Reply:
wdym ? it's just code for draining `fifo` should it still contains elements once we exit the main loop, so it will contains reductions